sherpa/astro/utils/tests/test_astro_utils_unit.py

```diff
@@ -220,3 +224,269 @@ def test_do_group_single_group(func, expected):
+    if data_class == "1d":
+        return Data1D('x1', x0, y)
+
+    if data_class == "1dint":
+        return Data1DInt('xint1', x0, np.asarray([3, 5, 8, 15]), y)
+
+    chans = np.arange(1, 5)
+    if data_class == "pha":
+        return DataPHA('pha', chans, y)
+
+    # These are a special case to check out grpuping and quality handling.
+    grp = np.asarray([1, -1, 1, 1])
+    qual = np.asarray([0, 0, 2, 0])
+    pha = DataPHA('pha', chans, y, grouping=grp, quality=qual)
+    if data_class == "grp":
+        return pha
+
+    if data_class == "qual":
+        pha.ignore_bad()
+        return pha
+
+    x0 = np.asarray([1, 2, 3] * 2)
+    x1 = np.asarray([1, 1, 1, 2, 2, 2])
+    y = np.asarray([2, 3, 4, 5, 6, 7])
+    if data_class == "2d":
+        return Data2D('x2', x0, x1, y, shape=(2, 3))
+
+    if data_class ==  "2dint":
+        return Data2DInt('xint2', x0, x1, x0 + 1, x1 + 1, y, shape=(2, 3))
+
+    if data_class == "img":
+        return DataIMG('img', x0, x1, y, shape=(2, 3))
+
+    if data_class == "imgint":
+        return DataIMGInt('imgi', x0, x1, x0 + 1, x1 + 1, y, shape=(2, 3))
+
+    assert False
+
+
+@pytest.mark.parametrize("data_class", ["1d", "1dint", "pha", "grp", "qual"])
+def test_calc_data_sum_invalid_range(data_class):
+    """lo > hi"""
+
+    data = make_data(data_class)
+    with pytest.raises(IOErr) as err:
+        utils.calc_data_sum(data, 10, 2)
+
+    # TODO: This error message is not great.
+    #
+    assert str(err.value) == "the energy range is not consistent, 10 !< 2"
+
+
+@pytest.mark.parametrize("data_class", ["1d", "1dint", "pha", "grp", "qual"])
+def test_calc_data_sum_no_range(data_class):
+    """Call calc_data_sum(data)"""
+
+    data = make_data(data_class)
+    assert utils.calc_data_sum(data) == 14
+
+
+@pytest.mark.parametrize("name", ["lo", "hi"])
+@pytest.mark.parametrize("limit,expected", [(0, 0), (1, 2), (2, 0), (3, 3),
+                                            (4, 0), (6, 0), (7, 4), (11, 0), (12, 5),
+                                            (13, 0), (15, 0)])
+def test_calc_data_sum_only_one_limit_1d(name, limit, expected):
+    """Call calc_data_sum(data, lo or hi): Data1D
+
+    It is not clear what the intended behavior is - either for the
+    PHA case, which seems to drive the calc_data_sum code, or for
+    other data sets, so just act as a regression test.
+    """
+
+    data = make_data("1d")
+    orig = data.get_dep(filter=True).copy()
+    assert utils.calc_data_sum(data, **{name: limit}) == expected
+    assert data.get_dep(filter=True) == pytest.approx(orig)
+
+
+@pytest.mark.parametrize("name", ["lo", "hi"])
+@pytest.mark.parametrize("limit,expected", [(0, 0), (1, 0), (2, 2), (3, 0),
+                                            (4, 3), (6, 0), (7, 0), (11, 0), (12, 0),
+                                            (13, 5), (15, 0)])
+def test_calc_data_sum_only_one_limit_1dint(name, limit, expected):
+    """Call calc_data_sum(data, lo or hi): Data1DInt"""
```

Comment:
```suggestion
    """Call calc_data_sum(data, lo or hi): Data1DInt and check that is does not mess up the internal state"""
```

Reply:
I don't understand the suggested change.

Reply:
Not the best wording. I meant to explain where there is the line `assert data.get_dep(filter=True) == pytest.approx(orig)`. It's not clear from the test that there is even the possibility that `calc_data_sum` messes up the dependent array, when then, when you look into the `calc_data_sum` code, there is actually a lot going on in the try-finally block. So, this check makes sense - it's just no obvious that it does. 

But I can live without that comment if we can't come up with better wording, not a point to break sweat about.